packages/search/src/search.tsx

```diff
@@ -90,349 +96,349 @@ type State = {
+		rangeRect.left;
+	return scrollLeft;
+};
 
-		if ( this.state.keyword === prevState.keyword ) {
-			return;
-		}
+type ImperativeHandle = {
+	focus: () => void;
+	blur: () => void;
+	clear: () => void;
+};
 
-		// if there's a keyword change: trigger search
-		if ( this.state.keyword ) {
-			this.onSearch( this.state.keyword );
+const InnerSearch: React.ForwardRefRenderFunction< ImperativeHandle, Props > = (
+	{
+		delaySearch: delaySearchProp,
+		disabled,
+		pinned,
+		onSearchClose,
+		onSearchChange,
+		onSearch: onSearchProp,
+		onBlur: onBlurProp,
+		onKeyDown: onKeyDownProp,
+		onClick,
+		describedBy,
+		delayTimeout,
+		defaultValue,
+		defaultIsOpen,
+		autoFocus,
+		onSearchOpen,
+		recordEvent,
+		overlayStyling,
+		placeholder: placeholderProp,
+		__,
+		inputLabel,
+		disableAutocorrect,
+		className,
+		dir,
+		fitsContainer,
+		searching,
+		compact,
+		hideOpenIcon,
+		openIconSide,
+		minLength,
+		maxLength,
+		hideClose,
+	},
+	forwardedRef
+) => {
+	const [ { keyword, isOpen, hasFocus }, setState ] = React.useReducer<
+		React.Reducer< State, Partial< State > >
+	>( ( oldState, nextState ) => ( { ...oldState, ...nextState } ), {
+		keyword: defaultValue ?? '',
+		isOpen: defaultIsOpen ?? false,
+		hasFocus: autoFocus ?? false,
+	} );
+
+	const instanceId = useInstanceId( InnerSearch, 'search' );
+	const searchInput: React.MutableRefObject< HTMLInputElement | null > = React.useRef( null );
+	const openIcon: React.RefObject< HTMLButtonElement | null > = React.useRef( null );
+	const overlay: React.RefObject< HTMLDivElement | null > = React.useRef( null );
+
+	const focus = React.useCallback( () => {
+		// if we call focus before the element has been entirely synced up with the DOM, we stand a decent chance of
+		// causing the browser to scroll somewhere odd. Instead, defer the focus until a future turn of the event loop.
+		setTimeout( () => searchInput.current?.focus(), 0 );
+	}, [] );
+
+	const blur = React.useCallback( () => searchInput.current?.blur(), [] );
+
+	const clear = React.useCallback( (): void => setState( { keyword: '' } ), [] );
+
+	React.useImperativeHandle(
+		forwardedRef,
+		() => ( {
+			focus,
+			blur,
+			clear,
+		} ),
+		[ focus, blur, clear ]
+	);
+
+	// uncontrolled
+	const delaySearch = React.useRef( delaySearchProp );
+
+	const doSearch: ( ( search: string ) => void ) & { cancel?: () => void } = React.useMemo(
+		() => ( delaySearch.current ? debounce( onSearchProp, delayTimeout ) : onSearchProp ),
+		[ onSearchProp, delayTimeout ]
+	);
+
+	useUpdateEffect( () => {
+		if ( keyword ) {
+			doSearch( keyword );
 		} else {
-			// explicitly bypass debouncing when there is no keyword is empty
-			if ( this.props.delaySearch ) {
+			// explicitly bypass debouncing when keyword is empty
+			if ( delaySearch.current ) {
 				// Cancel any pending debounce
-				this.onSearch.cancel?.();
+				doSearch.cancel?.();
 			}
 
-			this.props.onSearch( this.state.keyword );
+			doSearch( keyword );
 		}
```

Comment:
Another way to bypass debouncing (i.e., not wait when the keyword becomes empty) is:
```js
doSearch( keyword );
if ( keyword === '' ) {
  doSearch.flush?.();
}
```